RecoMuon/L2MuonSeedGenerator/src/Phase2L2MuonSeedCreator.cc

```diff
@@ -0,0 +1,842 @@
+              trajectoryStateTransform::persistentState(newTSOS, newTSOSDet->geographicalId().rawId());
+
+          // Emplace seed in output
+          LogDebug(metname) << "Emplacing seed in output";
+          output->emplace_back(L2MuonTrajectorySeed(seedTSOS, container, alongMomentum, l1TkMuRef));
+        }
+      }
+    }  // End seed emplacing (one seed per L1TkMu)
+  }  // End loop on L1TkMu
+  LogDebug(metname) << "All L1TkMu in event processed";
+  iEvent.put(std::move(output));
+}
+
+// In DT station 4 the top and bottom sectors are made of two chambers
+// due to material requirements. Online is not split:
+// Online sector 4 == offline sector 4 or 10, Online sector 10 == offline sector 10 or 14
+const std::vector<DTChamberId> Phase2L2MuonSeedCreator::matchingIds(const DTChamberId& stubId) const {
+  std::vector<DTChamberId> matchingDtIds;
+  matchingDtIds.reserve(2);
+  matchingDtIds.push_back(stubId);
+  if (stubId.station() == 4) {
+    if (stubId.sector() == 4) {
+      matchingDtIds.emplace_back(DTChamberId(stubId.wheel(), stubId.station(), 13));
+    }
+    if (stubId.sector() == 10) {
+      matchingDtIds.emplace_back(DTChamberId(stubId.wheel(), stubId.station(), 14));
+    }
+  }
+  return matchingDtIds;
+}
+
+// Pair bestSegIndex, quality for DT segments matching
+const std::pair<int, int> Phase2L2MuonSeedCreator::matchingStubSegment(const DTChamberId& stubId,
+                                                                       const l1t::MuonStubRef stub,
+                                                                       const DTRecSegment4DCollection& segments,
+                                                                       const l1t::TrackerMuonRef l1TkMuRef) const {
+  const std::string metname = "RecoMuon|Phase2L2MuonSeedCreator";
+
+  int bestSegIndex = -1;
+  int quality = -1;
+  unsigned int nHitsPhiBest = 0;
+  unsigned int nHitsThetaBest = 0;
+
+  LogDebug(metname) << "Matching stub with DT segment";
+  int nMatchingIds = 0;
+
+  for (DTChamberId id : matchingIds(stubId)) {
+    DTRecSegment4DCollection::range segmentsInChamber = segments.get(id);
+    for (DTRecSegment4DCollection::const_iterator segment = segmentsInChamber.first;
+         segment != segmentsInChamber.second;
+         ++segment) {
+      ++nMatchingIds;
+      DTChamberId segId = segment->chamberId();
+      LogDebug(metname) << "Segment DT detId: " << segId << ". RawId: " << segId.rawId();
+
+      // Global position of the segment
+      GlobalPoint segPos = dtGeometry_->idToDet(segId)->toGlobal(segment->localPosition());
+
+      // Check delta phi
+      double deltaPhi = std::abs(segPos.phi() - stub->offline_coord1());
+      LogDebug(metname) << "deltaPhi: " << deltaPhi;
+
+      double deltaTheta = std::abs(segPos.theta() - 2 * std::atan(std::exp(-l1TkMuRef->phEta())));
+      LogDebug(metname) << "deltaTheta: " << deltaTheta;
+
+      if (deltaPhi > matchingPhiWindow_ or deltaTheta > 0.4) {
```

Comment:
Rough check in delta theta, would it be better if it were something like `4 * matchingThetaWindow_`?

Reply:
I was more wondering where the 0.4 comes from and if it should be a named constants.

Reply:
The module looks for a match by checking, in order, delta phi, hit multiplicity and delta theta. The checks following the first are basically used to pick the highest-quality match, but I've resorted to have a rough check in eta anyways to avoid bad matches if nothing better is found. Since the current theta window is set to 0.1, for the rough check I've decided to go for something quite a bit larger (0.4 in this case), but there's no specific reason for that particular value

Reply:
As I see 0.4 repeated several times throughout the code of this class, I would prefer if you could make it a named constant and put a comment about it for later maintainability.